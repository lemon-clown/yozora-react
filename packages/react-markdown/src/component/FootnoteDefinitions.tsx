import { css, cx } from '@emotion/css'
import type { FootnoteDefinition } from '@yozora/ast'
import { useNodeRendererContext } from '@yozora/core-react-renderer'
import { tokens } from '@yozora/core-react-theme'
import React from 'react'
import { FootnoteDefinitionRenderer } from './renderer/footnoteDefinition'

export interface IFootnoteDefinitionsProps {
  /**
   * Title of the footnote definitions.
   */
  footnoteDefinitionsTitle?: React.ReactNode
  /**
   * If true, then the footnote definitions wont be render.
   */
  dontNeedFootnoteDefinitions?: boolean
}

export const FootnoteDefinitions: React.FC<IFootnoteDefinitionsProps> = props => {
  const { footnoteDefinitionsTitle, dontNeedFootnoteDefinitions = false } = props
  const { footnoteDefinitionMap } = useNodeRendererContext()
  const children = React.useMemo<React.ReactNode>(() => {
    // Get all of footnote reference definitions.
    const footnoteDefinitions: ReadonlyArray<FootnoteDefinition> =
      Object.values(footnoteDefinitionMap)
    if (footnoteDefinitions.length <= 0) return null

    return footnoteDefinitions.map(item => (
      <li key={item.identifier} id={item.identifier}>
        <FootnoteDefinitionRenderer {...item} />
      </li>
    ))
  }, [footnoteDefinitionMap])

  if (dontNeedFootnoteDefinitions || children === null) return null

  return (
    <div className={cx('yozora-footnote-definitions', classes.container)}>
      <div className={classes.title}>{footnoteDefinitionsTitle}</div>
      <ul className={classes.main}>{children}</ul>
    </div>
  )
}

const classes = {
  container: css({
    marginTop: '2rem',
    fontSize: '0.8rem',
  }),
  title: css({
    padding: 0,
    borderBottom: `1px solid ${tokens.colorBorderHeading}`,
    margin: '0 0 1rem',
    fontStyle: 'italic',
  }),
  main: css({
    padding: 0,
    margin: 0,
    listStyle: 'none',
  }),
}
