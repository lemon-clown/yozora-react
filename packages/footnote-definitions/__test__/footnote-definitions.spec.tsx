import { render } from 'enzyme'
import React from 'react'
import FootnoteDefinitions from '../src'
import type { IFootnoteItem } from '../src'

const nodes: IFootnoteItem[] = [
  {
    label: '1',
    identifier: 'footnote-1',
    children: <span>waw</span>,
  },
  {
    label: '2',
    identifier: 'footnote-2',
    children: <span>waw2</span>,
  },
]

describe('prop types', function () {
  beforeEach(() => {
    jest.spyOn(global.console, 'error').mockImplementation((...args) => {
      throw new Error(args.join(' '))
    })
  })

  describe('className is optional', function () {
    it('default', function () {
      const node = render(<FootnoteDefinitions nodes={nodes} />)
      expect(node.hasClass('yozora-footnote-definitions')).toBeTruthy()
    })

    it('custom', function () {
      const node = render(<FootnoteDefinitions nodes={nodes} className="my-footnote-definitions" />)
      expect(node.hasClass('yozora-footnote-definitions')).toBeTruthy()
      expect(node.hasClass('my-footnote-definitions')).toBeTruthy()
    })
  })

  it('style is optional', function () {
    const node = render(<FootnoteDefinitions nodes={nodes} style={{ color: 'orange' }} />)
    expect(node.css('color')).toEqual('orange')
  })
})

describe('snapshot', function () {
  it('basic', function () {
    const wrapper = render(
      <FootnoteDefinitions
        nodes={nodes}
        className="custom-footnote-definitions"
        style={{ marginTop: '2rem' }}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
