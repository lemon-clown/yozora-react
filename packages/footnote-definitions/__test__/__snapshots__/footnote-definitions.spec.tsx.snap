// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot basic 1`] = `
.c0 {
  overflow-x: auto;
  margin-top: 2rem;
  font-size: 0.8rem;
}

.c0 > .yozora-footnote-definitions__title {
  padding: 0;
  border-bottom: 1px solid var(--color-border-secondary);
  margin: 0 0 1rem;
  font-style: italic;
}

.c0 > .yozora-footnote-definitions__main {
  padding: 0;
  margin: 0;
  list-style: none;
}

.c0 .yozora-footnote-definitions__item {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
  padding: 0;
  margin: 0;
}

.c0 .yozora-footnote-definitions__item .yozora-footnote-definitions__item-title {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c0 .yozora-footnote-definitions__item .yozora-footnote-definitions__item-content {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: inline-block;
  margin: 0;
  padding: 0;
}

<div
  class="c0 yozora-footnote-definitions custom-footnote-definitions"
  style="margin-top:2rem"
>
  <div
    class="yozora-footnote-definitions__title"
  >
    footnote-definitions
  </div>
  <ul
    class="yozora-footnote-definitions__main"
  >
    <li
      class="yozora-footnote-definitions__item"
      id="footnote-1"
    >
      <p
        class="yozora-footnote-definitions__item-title yozora-paragraph"
      >
        <a
          href="#reference-footnote-1"
        >
          ↑
        </a>
          [1]:   
      </p>
      <div
        class="yozora-footnote-definitions__item-content"
      >
        <span>
          waw
        </span>
      </div>
    </li>
    <li
      class="yozora-footnote-definitions__item"
      id="footnote-2"
    >
      <p
        class="yozora-footnote-definitions__item-title yozora-paragraph"
      >
        <a
          href="#reference-footnote-2"
        >
          ↑
        </a>
          [2]:   
      </p>
      <div
        class="yozora-footnote-definitions__item-content"
      >
        <span>
          waw2
        </span>
      </div>
    </li>
  </ul>
</div>
`;
